New Plugin for Nod SDK v1686

Khwaab opened the pull request:
New plugin which is compatible with new SDK and backspin. Moved old plugin to OLD directory. Updated Readme.

2nd commit adds some fixes and enables multiple devices.

Khwaab committed:
New Plugin for Nod SDK v1686

New plugin which is compatible with new SDK and backspin. Updated Readme.

Khwaab committed:
Fixes + Multiple devices

Fix hardcoding of device left from debugging.

Add ability to connect to multiple devices. Each device will be
accessible by a path corresponding to its id.
i.e. NodOSVRPlugin/0/interface, for the first device
NodOSVRPlugin/1/interface, for the second

Fix OpenSpatial Dll to send the id of the device, and use that id
to update the appropriate state.

rpavlik commented on NodOSVR.json:
```diff
@@ -0,0 +1,68 @@
+    },
+    "button": {
+      "count": 10,
+      "traits": [
```
Buttons don't have traits - you might try using the web tool to edit this: http://tools.getosvr.org/json-editor/

rpavlik commented on NodOSVR.json:
```diff
@@ -0,0 +1,68 @@
+    }
+  },
+  "semantic": {
+    "control": {
```
This layer of nesting (with `control` and `track`) is probably unnecessary - I'd put the contents of `control` at the top level

rpavlik commented on NodOSVR.json:
```diff
@@ -0,0 +1,68 @@
+    },
+    "track": {
+      "hands/right" : "tracker"
+    }
```
instead of what you have in `track` just add in at the top level something like `"orientation": "tracker/0"` since what you have right now may or may not work but isn't ideal. (The device isn't always on the right hand, etc.

rpavlik replied:
You might also be able to do `"$target": "tracker/0"` which would make the parent path (that is, `..../semantic` itself) be the tracker, which would work nicely with the `*` below.

rpavlik commented on NodOSVR.json:
```diff
@@ -0,0 +1,68 @@
+  },
+  "automaticAliases": {
+    "/controller": "semantic/control",
+    "/me" :  "semantic/track"
```
I would not create this automatic alias.

rpavlik commented on NodOSVR.json:
```diff
@@ -0,0 +1,68 @@
+    }
+  },
+  "automaticAliases": {
+    "/controller": "semantic/control",
```
You might instead suggest this automatic alias line:

``` json
"/controller": "semantic/*"
```

which will recursively alias the tree under `/controller`.

rpavlik commented on NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,210 @@
+
+// Global device count for path incrementation
+int deviceCount = 0;
+std::vector<Backspin*> devices;
```
The count is redundant, I would think.

rpavlik replied:
Oh, and surprised this compiles, since you don't have a declaration of `class Backspin` visible at this point.

Khwaab replied:
Oh, yeah, I fixed that must not have pushed up my latest changes, thanks.

rpavlik commented on NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        // send orientation quaternion and 0,0,0 for position (could just send orientation, but
+        // pose is expected by most programs
```
Only send what you actually have - the mapping is the core's responsibility.

Khwaab replied:
Are you talking about sending 0 for the position of the pose? The reason I'm doing this is just because our "pose" tracking is just finishing up and I will be implementing it very soon. So I'm just putting a placeholder here for now.

rpavlik commented on NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,210 @@
+        osvrDeviceTrackerSendPose(m_dev, m_orientation, &pose, 0);
+
+        return OSVR_RETURN_SUCCESS;
+    }
```
So the typical model for an object-based async device like this is that you have a blocking call in the update method. Since that doesn't seem to be the case, you'll probably be getting race conditions depending on how you're updating those values, since the update method gets its own thread to run in.

You might want to either restructure your use of your SDK, if possible (moving responsibility into the individual objects), or use a simpler, "dumb" struct to store the per-device OSVR objects, handling it from your other code (that I haven't gotten to reading yet)

rpavlik commented on NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,210 @@
+    HardwareDetection() : m_found(false) { NodInitialize(OpenSpatialEventFired); }
+    OSVR_ReturnCode operator()(OSVR_PluginRegContext ctx) {
+        std::cout << "[NOD PLUGIN] Got a hardware detection request" << std::endl;
+        if (!m_found)
```
You don't need this `if (!m_found)` if you have a better way of keeping track of if you're already handling a device.

rpavlik commented on NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,210 @@
+        + sinHalfRoll * sinHalfPitch * sinHalfYaw;
+}
+
+void OpenSpatialEventFired(NodEvent ev)
```
Where does this get called? (in what thread, etc.)? It looks like there aren't any calls to transfer control to your SDK, suggesting that these event handlers happen in another thread (and thus need thread-safety, etc.) Now, I think you can use an async device token (without an update callback registered, even, I think) to send data from any thread, not just the one that OSVR will spawn for that device, but it wasn't necessarily a designed-for use case.

Also, it's unusual to have a callback that doesn't have a `void *` parameter for arbitrary user data (usually a pointer to data) to be passed it, and makes your API much harder to use.

Khwaab replied:
Thanks for the void\* comment that's a good point.

So yeah it is running in a different thread. Haven't implemented full thread safety yet it'll be coming soon though. 

So If I understand you correctly you are saying instead of registering the update callback and storing / polling values updated by our SDK, I should be able to call the set value functions from our SDK thread? If so, awesome, I was attempting to do that before and was running into some issues, but I would definitely prefer that approach, I'll look into it further for sure. Thanks

Khwaab committed:
Update Plugin to Async Notification Based Model

1. No longer poll in an update method but just send data immidiately
when received from Nod callback
2. Make sure multiple devices can be accessed seprately and be identified
3. Add a client example to show how plugin is used in C++

NOTE: No fixes were made to the json, nor was a example server_config made,
we are still exploring how we want to map the backspin into OSVR's exposed
environment. However, this plugin will enable devlopers to start using Nod
data through the OSVR environment.

Khwaab committed:
Update Json Clean Up Directories

gfrolov commented on Plugin/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+# This generates a header file, from the named json file, containing a string literal
+# named com_osvr_example_selfcontained_json (not null terminated)
+# The file must be added as a source file to some target (as below) to be generated.
+MESSAGE("converting")
```
This message is not needed. Please remove

gfrolov commented on Plugin/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+    NodOSVRPlugin.cpp
+    "${CMAKE_CURRENT_BINARY_DIR}/NodOSVR.h"
+	Settings.h
+	NodPlugin.h)
```
Settings.h and NodPlugin.h shouldn't be included explicitly since they are part of the OpenSpatial SDK. We should add findOpenSpatial.cmake to find the required include folder and libraries

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+    <http://sensics.com/osvr>
+*/
+
+// Copyright 2014 Sensics, Inc.
```
Please change the copyright notices accordingly

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+/** @file
+    @brief Comprehensive example: Implementation of a dummy Hardware Detect
+   Callback that creates a dummy device when it is "detected"
```
I think this comment can be removed

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+#include "NodOSVR.h"
+#include "NodPlugin.h"
+
+#pragma comment(lib, "OpenSpatialDll.lib")
```
Similar to my comment in CMakelists.txt. We should really be finding OpenSpatial SDK location and linking them from there, especially since there are 32 and 64 bit versions available.

gfrolov commented on Plugin/NodOSVR.json:
```diff
@@ -0,0 +1,58 @@
+      ]
+    },
+    "button": {
+      "count": 10,
```
Remove the comma which makes the JSON file invalid

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+#include <math.h>
+
+// Global device count for path incrementation
+int deviceCount = 0;
```
I think deviceCount is not really needed here. You should just be able to use the size of devices vector instead

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+        OSVR_Pose3 pose;
+        pose.rotation = orientationval;
+        //TODO: when 6dof is implemented send translations
+        pose.translation = { 0,0,0 };
```
I would comment this out because like Ryan mentioned, Core will take care of it and this is unnecessary

gfrolov commented on .clang-format:
```diff
@@ -1,67 +0,0 @@
----
```
We should leave this file

gfrolov commented on Example Client/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+# autodetected but you might need to specify osvr_DIR to be something like
+# C:/Users/Ryan/Desktop/build/OSVR-Core-vc12 or
+# C:/Users/Ryan/Downloads/OSVR-Core-Snapshot-v0.1-406-gaa55515-build54-vs12-32bit
+# in the CMake GUI or command line
```
Can remove these comments

gfrolov commented on Example Client/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+find_package(osvr)
+
+# Pass as many source files as you need. You'll want your own since this just uses
+# the regular Tracker callback demo source file (to avoid duplicate sources).
```
Can remove these comments as well

gfrolov commented on Example Client/TrackerCallback.cpp:
```diff
@@ -0,0 +1,116 @@
+    <http://sensics.com/osvr>
+*/
+
+// Copyright 2014 Sensics, Inc.
```
Please change the copyright notices accordingly

gfrolov commented on osvr_server_config.nod.sample.json:
```diff
@@ -1,16 +0,0 @@
-{
```
We still need the osvr_server_config file. So don't delete it

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+    if (ev.type == EventType::EulerAngles)
+    {
+        Backspin* dev = devices.at(ev.sender);
+        double val[4];
```
Can't we just use OSVR_OrientationState type instead of double arr, so we don't need to copy the values to another variable?

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+// Anonymous namespace to avoid symbol collision
+namespace {
+
+void euler(const float& pitch, const float& roll, const float& yaw, double* mData)
```
This could either return `OSVR_OrientationState` or pass `OSVR_OrientationState* mData` instead of `double* mData`.

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+    const float sinHalfRoll = (float)sin(roll / 2.0f);
+    const float cosHalfRoll = (float)cos(roll / 2.0f);
+
+    mData[0] = cosHalfRoll * cosHalfPitch * cosHalfYaw
```
Please add some comments on what each formula calculates or explicitly name contents of quaternion because right now it hard to say which value is W, X, Y or Z.

gfrolov commented on Plugin/NodOSVRPlugin.cpp:
```diff
@@ -0,0 +1,180 @@
+        double val[4];
+        euler(-ev.roll, -ev.pitch, ev.yaw, val);
+        OSVR_OrientationState orientationval;
+        memcpy(orientationval.data, val, 4);
```
Verify that your euler function returns values in same order as `OSVR_Quaternion` expects it to be (it's not clear right now from the code) 
`OSVR_Quaternion` struct contents are `W, X, Y, Z`

Khwaab commented:
There is another commit which is merged into my master branch, but is somehow not reflected here. Maybe it just takes some time, but if it doesn't take I may close this one and open a new one.

gfrolov commented:
The main Nod plugin repo was changed from private to public so your fork remained private and might have been locked if you were using a free account. More info here - https://help.github.com/articles/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility/#changing-a-private-repository-to-a-public-repository 
You will need to change it to public to unlock https://help.github.com/articles/making-a-private-repository-public/

Khwaab commented:
I unlocked my repository and then pushed my changes, however they aren't
reflected in the pull request

Might just open a new one.
On Wednesday, November 4, 2015, Georgiy Frolov notifications@github.com
wrote:

> The main Nod plugin repo was changed from private to public so your fork
> remained private and might have been locked if you were using a free
> account. More info here -
> https://help.github.com/articles/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility/#changing-a-private-repository-to-a-public-repository
> You will need to change it to public to unlock
> https://help.github.com/articles/making-a-private-repository-public/
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OSVR/OSVR-Nod/pull/1#issuecomment-153723319.

rpavlik commented:
Yep, that is the way to go - notice that this repo (as of this writing) says it has 0 forks. Changing from private to public breaks the repo network in GitHub's system.

Khwaab commented:
Starting a new pull request (this one isn't updating)
